Modules/FindCUDA/select_compute_arch.cmake

```diff
@@ -0,0 +1,175 @@
+
+if (CUDA_VERSION VERSION_GREATER "6.5")
+  set(CUDA_KNOWN_GPU_ARCHITECTURES "3.2 3.7 ${CUDA_KNOWN_GPU_ARCHITECTURES} 5.2")
+  message("CUDA_VERSION VERSION_GREATER 6.5 YES ARCH=${CUDA_KNOWN_GPU_ARCHITECTURES}")
+endif ()
+
+if (CUDA_VERSION VERSION_GREATER "7.5")
+  set(CUDA_KNOWN_GPU_ARCHITECTURES "${CUDA_KNOWN_GPU_ARCHITECTURES} 6.0 6.2")
+  message("CUDA_VERSION VERSION_GREATER 7.5 YES ARCH=${CUDA_KNOWN_GPU_ARCHITECTURES}")
+endif ()
+
+
+################################################################################################
+# A function for automatic detection of GPUs installed  (IF autodetection is enabled)
+# Usage:
+#   CUDA_DETECT_INSTALLED_GPUS(OUT_VARIABLE)
+function(CUDA_DETECT_INSTALLED_GPUS OUT_VARIABLE)
+  if(NOT CUDA_GPU_DETECT_OUTPUT)
+    set(cufile ${PROJECT_BINARY_DIR}/detect_cuda_archs.cu)
+
+    file(WRITE ${cufile} ""
+      "#include <cstdio>\n"
+      "int main()\n"
+      "{\n"
+      "  int count = 0;\n"
+      "  if (cudaSuccess != cudaGetDeviceCount(&count)) return -1;\n"
+      "  if (count == 0) return -1;\n"
+      "  for (int device = 0; device < count; ++device)\n"
+      "  {\n"
+      "    cudaDeviceProp prop;\n"
+      "    if (cudaSuccess == cudaGetDeviceProperties(&prop, device))\n"
+      "      std::printf(\"%d.%d \", prop.major, prop.minor);\n"
+      "  }\n"
+      "  return 0;\n"
+      "}\n")
+
+    execute_process(COMMAND "${CUDA_NVCC_EXECUTABLE}" "--run" "${cufile}"
+                    WORKING_DIRECTORY "${PROJECT_BINARY_DIR}/CMakeFiles/"
+                    RESULT_VARIABLE nvcc_res OUTPUT_VARIABLE nvcc_out
+                    ERROR_QUIET OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+    if(nvcc_res EQUAL 0)
+      string(REPLACE "2.1" "2.1(2.0)" nvcc_out "${nvcc_out}")
+      set(CUDA_GPU_DETECT_OUTPUT ${nvcc_out} CACHE INTERNAL "Returned GPU architetures from detect_gpus tool" FORCE)
+    endif()
+  endif()
+
+  if(NOT CUDA_GPU_DETECT_OUTPUT)
+    message(STATUS "Automatic GPU detection failed. Building for all known architectures.")
+    set(${OUT_VARIABLE} ${CUDA_KNOWN_GPU_ARCHITECTURES} PARENT_SCOPE)
+  else()
+    set(${OUT_VARIABLE} ${CUDA_GPU_DETECT_OUTPUT} PARENT_SCOPE)
+  endif()
+endfunction()
+
+
+################################################################################################
+# Function for selecting GPU arch flags for nvcc based on CUDA_ARCH_NAME
+# Usage:
+#   SELECT_NVCC_ARCH_FLAGS(out_variable)
+function(CUDA_SELECT_NVCC_ARCH_FLAGS out_variable)
+  # List of arch names
+# CUDA_ARCH_NAME: One of:
+  set(archs_names "Fermi" "Kepler" "Kepler-M" "Tesla" "Maxwell" "Maxwell-M" "Pascal" "All" "Manual")
+  set(archs_name_default "All")
+  if(NOT CMAKE_CROSSCOMPILING)
+    list(APPEND archs_names "Auto")
+    set(archs_name_default "Auto")
+  endif()
+
+  # SET CUDA_ARCH_NAME strings (so it will be seen as dropbox in CMake-Gui)
+  set(CUDA_ARCH_NAME ${archs_name_default} CACHE STRING "Select target NVIDIA GPU achitecture.")
```

Comment:
Note that I tried really hard to make all the "public" variables for CUDA be lists instead of strings.  I think for consistency we should do the same thing.  There are CMake functions for converting lists to strings, so you wouldn't have to change your code here much.